[Controllers] Restructure protocolIndex / controllerIndex and cplugin_ID

TD-er opened the pull request:
This was rather messy code and no checks.
It is now somewhat similar to the code for plugins defining the relation between:
- Plugin ID
- Task Index
- Device Index.

Also added the option to call CPLUGIN_EXIT and a call for 10/sec loops which we apparently didn't have until now for controllers.

Fixes: #2821
Fixes: #2866

TD-er committed:
Revert "[WiFi] Improve reconnect management"

This reverts commit 91b7b3b9a2c1de78073dc046983ca3877f3c3cb5.

TD-er committed:
[Controllers] Restructure protocolIndex / controllerIndex and cplugin_ID

This was rather messy code and no checks.
It is now somewhat similar to the code for plugins defining the relation between:
- Plugin ID
- Task Index
- Device Index.

TD-er committed:
[RN2483] Add option to reset LoRa module using reset pin

Still not working well on RN2483 with firmware 1.0.1 and a slow rising voltage at power on.

TD-er committed:
[Build] Add compile time checks to make sure offsets in settings are OK

Compute some offsets of members in settings struct and security struct to make sure padding doesn't differ among builds on different platforms.

TD-er commented:
I made a [test build](https://www.dropbox.com/s/jjaoos32a12werb/ESPEasy_mega-20191208-36-PR_2864.zip?dl=0)

Would be appreciated if someone could test it with respect to controllers.
Especially:
- Do settings work well when switching to/from this build? (save controller settings with this build active)
- Any issues with controllers? (e.g. missing credentials, controllers not connecting, etc.)
- Any issues with plugins? (enable/disable plugins or controllers and test it)
- Switch between builds where some plugins or controllers are not available (e.g. switch between "normal" and "test")

N.B. the bug when using static IP is still not fixed.   (will crash)

v-a-d-e-r commented:
I merged your changes with the controller stuff to my ESPeasy. P2P and MQTT seem to work. But the loop counter droped about 800-1000! My version: ~4500, my version with your changes: ~3700. What take so much time now?

TD-er commented:
> I merged your changes with the controller stuff to my ESPeasy. P2P and MQTT seem to work. But the loop counter droped about 800-1000! My version: ~4500, my version with your changes: ~3700. What take so much time now?

Thanks for testing.
I assume you're not using the "eco mode" otherwise it would not have made any difference as then the idle time forces the loop counter to only what is needed.

I will have a look at the code to see if I can explain it.
I did add some more checks, but on the other hand the lookup should have been taking less.
At least it is something to check as it may indicate some unneeded extra work.
Nice catch.

TD-er commented:
I've looked at the code and it does look like the check for `validPluginID` is a bit more CPU intensive right now and that's also used quite often.
For the controllers, the same function is also as much more resource hungry, but controllers are not that actively polled.

I did add a 10/sec call for controllers, which does call `getProtocolIndex`. That's also performing an extra search that could be done more elegant I guess.

TD-er committed:
[Controller] Optimization by reducing number of checks.

As discussed [here](https://github.com/letscontrolit/ESPEasy/pull/2864#issuecomment-580894393)

v-a-d-e-r commented:
Ok, I merged the new changes, but I don't wanna say it: The loop counter droped again here about ~900 to ~2800. 8-/
And yes, eco mode is OFF for testing purposes....
CPU load: 4%, LC: 4511 
CPU load: 4%, LC: 3864 (with the new controller stuff)
CPU load: 4%, LC: 2702 (with recent changes)

TD-er commented:
Well the CPU load is roughly the same, which indicates the amount of work to be done during "active" loops takes the same amount of time.
So this means the changes do have impact on the "idle loops" so they do take longer to process.

I am not entirely sure what impact it may have in the end as lower loop count doesn't mean it is slower.
To understand it, I will try to explain it without going into too much detail.

What happens:
- Active loops are loops where the scheduler is performing a single scheduled action.
- Inactive loops are loops where the scheduler does not have something ready to be executed.

The CPU load is (for over a year now) computed by the amount of time spent in active and inactive loops.
So if the CPU load remains the same, this means it takes the same amount of time for each "active" loop per second. (actually per minute interval)

The loop count (LC) is the number of times the `loop()` function is called per second.
The max is roughly 15k - 22k. depending on the core version.
In "Eco" mode, the idle loops are only calling `delay()` as much as possible, so the loop count is a lot lower to only have active loops and the inactive loop takes exactly as long as needed to get to the next active loop. (with some upper limit in delay time to keep the unit responsive)

Regardless of eco mode enabled or not, during the inactive loop there is also a call to perform the background tasks.
These are about handling incoming traffic, handling the MQTT broker connection, etc.
If for whatever reason the interval between calling these background tasks is longer, then a single call to the background task can take longer as there is more work to be done.

Now the paradoxical situation where a lower loop call can yield a faster node (not saying it is, but it is possible).
If the idle loops are more interleaved with the active calls, then there is more time between "idle loops".
For example:
- Active loop: 1,3,5,7
- Inactive loop: 2,4,6,8

This has a longer time between inactive loops compared to:
- Active loop: 1,2, - , 6,7
- Inactive loop: 3,4,5, - , 8,9,10

The last situation has a higher loop count, but in inactive loops 4,5 & 9,10 its call to the background tasks will return almost immediately.
On the other hand, the time between idle loop 3 and 8 is longer, so the unit may respond slower to external activity like network traffic and sending out data from controllers.


The things changed in this PR is having a lot more structure and checks on the internal administration of present controllers and plugins.
So I would expect a bit more CPU load as it does have more checks than before.
On the other hand finding the right plugin or controller may take less time. So in the end it should be roughly the same.

In the last commit I made yesterday I removed some extensive checks which are not always needed.
So the CPU load should be lower (although it is hard to predict how much).
But on the other hand the internal administration of the CPU load may be less accurate when switching a lot between active and inactive loops.
If you are perfect alternating between active and inactive loops, the internal administration is having the most overhead, so that can also skew the numbers a bit.

TD-er commented:
Just for reference on a module that was using this PR:

Before last commit | -
-- | --
Uptime: | 0 days 0 hours 4 minutes
Load: | 17.70% (LC=2394)

With last commit | -
-- | --
Uptime: | 0 days 0 hours 5 minutes
Load: | 17.90% (LC=2444)


I restarted it to make sure it was not "settled" for hours and also waited a few minutes so the data is not right after the reboot.

v-a-d-e-r commented:
Ok, thanks for the detailed explanation. I have tested it with enabled P2P and MQTT controller, 1 task (P001) and eco mode 'off'. I will make further tests to see what's going on with my test node and with different settings.

TD-er commented:
Don't focus too much on the loopcount, but also try to see if things are working stable and maybe also some latency timings to see how quick the node reacts to network activity.

v-a-d-e-r commented:
During my tests I found another issue with Eastron (P078) plugin:
1. If the serial communication is disconnected, the CPU load rises up over 66% (Normal: 8%)!
2. In "case PLUGIN_GET_DEVICEGPIONAMES:" is "serialHelper_getGpioNames(event);", but there is no function for this call!? Shouldn't it be "serialHelper_getGpioNames(event, true, true)"?

TD-er commented:
See the function declaration in the .h file (ESPEasy_fwdecl.h):
```c++
void     serialHelper_getGpioNames(struct EventStruct *event,
                                   bool                rxOptional = false,
                                   bool                txOptional = false);
```

The function declaration has default values for the last two values.
Simply stated, if the compiler agrees, then there is a function declaration that matches :)

Not only the Eastron plugin, but also other plugins using serial will keep the unit extremely occupied when the stated sensor is not connected.
The serial object does have a timeout setting which we can set more often to mitigate this problem.
Also I think we should have a look at more plugins to make them less blocking while waiting for values.
For example the SenseAir plugin (and other plugins using Modbus RTU) does have blocking calls as long as the sensor needs to give an answer.  On the SenseAir S8 this takes about 200 msec for each reading. (S11 does respond a lot faster)

The new controller I am working on (RN2483 LoRa) is even worse and may block the unit for a very long time when the module is not responding. Also during transfers it is blocking for roughly 2 seconds.


So yes, you are totally right about those and it is up for some serious code changes.
But I think/hope this behavior has not been introduced by the code change of this PR, correct?

v-a-d-e-r commented:
Right, it has nothing to do with this PR. I'm merging some other stuff and I found this issue by the way and was surprised about the compiler error about "too few arguments". But it was very fast fixed... ;-)

v-a-d-e-r commented:
Found a typo in Misc.ino & Globals.h:
#define DEFAULT_MQTT_USE_UNITNANE_AS_CLIENTID
=> UNITNAME

TD-er commented:
> DEFAULT_MQTT_USE_UNITNANE_AS_CLIENTID

Can't find anything with "nane", even when searching case insensitive and not on complete words.

TD-er committed:
[PIO] Update to core 2.6.3

v-a-d-e-r commented:
Found it in the last official version from 08.12.2019. Is it fixed meanwhile?

TD-er commented:
Apparently ;)  (but maybe not yet merged?)

TD-er committed:
[WIFi] Fix crashes with fixed IP

v-a-d-e-r commented:
Yes, can be I have missed it in the past to merge from a commit. ;)

TD-er committed:
Fix unexpected string padding in parseTemplate (#2866)

The extra padding to a minimal line length does only make sense for use on displays and some special formatting.
So the majority of the uses of parseTemplate now uses 0 as minimum line length.

Fixes: #2866
Closes: #2867 (superseeds PR)
